### PR TITLE
grafana add summary visualization

### PR DESCRIPTION
**Pull request by beilineili**
solving issue #68 
An interface /podsummary was added to grafanadi to display the pod status summary in grafana. 

At the beginning, the SLO timeout time (service 8ms, job 1m30s) is displayed. When the SLO timeout occurs, the error summary extracted from key lifecycle events is displayed.

image

image

image

### CODE REVIEW

File: internal/grafanadi/handler/podsummary_handler.go [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+	if handler.request.Method == http.MethodGet {
+		params.PodUIDName = handler.request.URL.Query().Get("searchkey")
+		params.PodUID = handler.request.URL.Query().Get("searchvalue")
+	}
+
+	handler.requestParams = &params
+	return nil
+}
+
+func (handler *PodSummaryHandler) ValidRequest() error {
```
Comment (hnhbwlp) [inline internal/grafanadi/handler/podsummary_handler.go:49]: Isn’t parameter verification required here?

File: internal/grafanadi/handler/podsummary_handler.go [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+		metrics.QueryMethodDurationMilliSeconds.WithLabelValues("QueryPodSummary").Observe(cost)
+	}()
+	util := interutils.Util{
+		Storage: handler.storage,
+	}
+	util.GetUid(podYamls, key, &value)
+	podUid := value
+
+	match, _ := regexp.MatchString("^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$", podUid)
+	if !match {
```
Comment (hnhbwlp) [inline internal/grafanadi/handler/podsummary_handler.go:75]: Is it necessary to reveal the error message to the user?

File: internal/grafanadi/handler/podsummary_handler.go (hunk 1 above)
Comment (beilineili) [reply internal/grafanadi/handler/podsummary_handler.go:49]: Thank you very much for your comment. I am writing code here to imitate the format of other existing files in the internal/gfanadi/handler project. Perhaps the author of this code has their own considerations? ```golang func (handler *PodPhaseHandler) ValidRequest() error {
	return nil
}` [https://github.com/alipay/container-observability-service/blob/container-lifecycle-summary-cross-arch/internal/grafanadi/handler/podphase_handler.go](url)

File: internal/grafanadi/handler/podsummary_handler.go (hunk 2 above)
Comment (beilineili) [reply internal/grafanadi/handler/podsummary_handler.go:75]: You are right, we can verify the correctness of "podUid" in other parts of the code without returning error messages to the user here. I will delete this code.

File: internal/grafanadi/handler/podsummary_handler.go [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+	}()
+	go func() {
+		defer wg.Done()
+		err = handler.storage.QuerySloTraceDataWithPodUID(&slotracedata, podUid)
+	}()
+	go func() {
+		defer wg.Done()
+		err = handler.storage.QueryLifePhaseWithPodUid(&lifephases, podUid)
+	}()
+	if err != nil {
```
Comment (linuzb) [inline internal/grafanadi/handler/podsummary_handler.go:88]:
```suggestion
	if err != nil {
```

Comment (beilineili) [reply internal/grafanadi/handler/podsummary_handler.go:88]: Assigning incorrect values to multiple go functions can indeed lead to inconsistent results. Therefore, I modified the code to set separate error variables for each go func, and then judged the errors after all go funcs were completed.
```golang
        go func() {
		defer wg.Done()
		podInfoErr = handler.storage.QueryPodInfoWithPodUid(&podInfos, podUid)
	}()
	go func() {
		defer wg.Done()
		sloTraceDataErr = handler.storage.QuerySloTraceDataWithPodUID(&slotracedata, podUid)
	}()
	go func() {
		defer wg.Done()
		lifePhaseErr = handler.storage.QueryLifePhaseWithPodUid(&lifephases, podUid)
	}()
        wg.Wait()
	if podInfoErr != nil {
		return http.StatusBadRequest, nil, errors.New(podInfoErr.Error())
	}
	if sloTraceDataErr != nil {
		return http.StatusBadRequest, nil, errors.New(sloTraceDataErr.Error())
	}
	if lifePhaseErr != nil {
		return http.StatusBadRequest, nil, errors.New(lifePhaseErr.Error())
	}
```

File: internal/grafanadi/handler/podsummary_handler.go [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+				}
+				if info["UserAgent"] != nil {
+					component = info["UserAgent"]
+				}
+			}
+			if obj, ok := info["eventObject"]; ok {
+				v, ok := obj.(map[string]interface{})
+				if ok {
+					msg := v["message"]
+					if len(msg.(string)) > 4 {
```
[the remark this answers is not in the document]
Comment (beilineili) [reply internal/grafanadi/handler/podsummary_handler.go:124]: Indeed, we can directly use the asserted result value, and there is no need to add a judgment here.

**Review by linuzb — APPROVED** [review]
LGTM

**Review by hnhbwlp — APPROVED** [review]
LGTM

**Pull request merged**